crates/next-api/src/module_graph.rs

```diff
@@ -716,46 +726,67 @@ impl GlobalBuildInformation {
     ) -> Result<Vc<ClientReferenceGraphResult>> {
         let span = tracing::info_span!("collect all client references for endpoint");
         async move {
-            let mut result = if let [graph] = &self.client_references[..] {
-                // Just a single graph, no need to merge results
-                graph
-                    .get_client_references_for_endpoint(entry)
-                    .owned()
-                    .await?
+            let result = if let [graph] = &self.client_references[..] {
+                // Just a single graph, no need to merge results  This also naturally aggregates
+                // server components and server utilities in the correct order
+                let result = graph.get_client_references_for_endpoint(entry);
+                #[cfg(debug_assertions)]
+                {
+                    let result = result.await?;
+                    if has_layout_segments {
+                        use rustc_hash::FxHashSet;
+
+                        let ServerEntries {
+                            server_utils,
+                            server_component_entries,
+                        } = &*find_server_entries(entry, include_traced).await?;
+                        // order of server utils doesn't matter, so just ensure that they match
+                        assert_eq!(
+                            FxHashSet::from_iter(result.server_utils.iter()),
+                            FxHashSet::from_iter(server_utils.iter())
+                        );
+                        // The order of server_components does matter, enforce it is identical
+                        assert_eq!(&result.server_component_entries, server_component_entries);
+                    }
+                }
+                result
             } else {
                 let results = self
                     .client_references
                     .iter()
                     .map(|graph| graph.get_client_references_for_endpoint(entry))
-                    .try_join()
-                    .await?;
+                    .try_join();
+                // Do this separately for now, because the aggregation of multiple graph traversals
+                // messes up the order of the server_component_entries.
+                let server_entries = async {
+                    if has_layout_segments {
+                        let server_entries = find_server_entries(entry, include_traced).await?;
+                        Ok(Some(server_entries))
+                    } else {
+                        Ok(None)
+                    }
+                };
+                let (results, server_entries) = join!(results, server_entries);
 
-                let mut iter = results.into_iter();
+                let mut iter = results?.into_iter();
                 let mut result = ReadRef::into_owned(iter.next().unwrap());
```

Comment:
Not sure if you'd want to use `ReadRef::into_owned` here. If this consumes the value, you'd have to recompute it next time around, which would be more expensive than cloning?

Reply:
Final read hint might be it:

https://github.com/vercel/next.js/blob/068ae14412e69feb2cde40bb222c45174f26a201/turbopack/crates/turbopack-ecmascript/src/lib.rs#L1834-L1848

> for a ReadRef im guessing this is basically always a clone since turbo tasks is holding a reference.

I.e. `into_owned` unnecessarily complicates the code here since it's always `clone` anyway? 🤷

Reply:
concur on complicated.  i do kinda hate the `(&*read_ref).clone()` pattern

however on second look this clone is unnecessarily expensive.  we only care to extract the client_references so cloning the rest of it is wasteful.